File: src/main/java/quantum/basis/CatBasis.java
package quantum.basis;

import org.ejml.data.ZMatrixRMaj;
import org.ejml.dense.row.CommonOps_ZDRM;

import quantum.io.MatrixIO;

public class CatBasis extends Basis {

    private final int N;    // Fock space dimension for encoding
    private final double[] alpha;

    public CatBasis(int dimension, int N, double alphaReal, double alphaImag) {
        this.dimension = dimension;
        this.N = N;
        this.alpha = new double[]{alphaReal, alphaImag};
        createBasis();
    }

    @Override
    public void createBasis() {
        StandardBasis fockStates = new StandardBasis(dimension);

        // Generate coherent state
        // Common factor
        double coherentFactor = Math.exp(-(Math.pow(alpha[0], 2) + Math.pow(alpha[1], 2)) / 2.0);
        long[] factorials = computeFactorials(N);
        double[] numerator;
        double denominator;
        double[] factor;

        ZMatrixRMaj coherentState = new ZMatrixRMaj(N, 1);
        for (int n = 0; n < N; n++) {
            numerator = complexPower(alpha, n);
            denominator = Math.sqrt(factorials[n]);
            factor = new double[]{numerator[0] / denominator, numerator[1] / denominator};

            ZMatrixRMaj temp = fockStates.getBasisStates()[n].copy();
            CommonOps_ZDRM.scale(factor[0], factor[1], temp);
            CommonOps_ZDRM.add(coherentState, temp, coherentState);
        }

        CommonOps_ZDRM.scale(coherentFactor, 1.0, coherentState);

        double[] phase;
        double[] phase_shift;
        basisStates = new ZMatrixRMaj[dimension];
        for (int ell = 0; ell < dimension; ell++) {
            phase_shift = new double[]{0.0, 0.0};
            basisStates[ell] = new ZMatrixRMaj(N, 1);
            for (int k = 0; k < N; k++) {
                phase = expI(2.0 * Math.PI * k / N * (N - ell));
                phase_shift[0] = phase_shift[0] + phase[0];
                phase_shift[1] = phase_shift[1] + phase[1];
            }

            basisStates[ell] = coherentState.copy();
            CommonOps_ZDRM.scale(phase_shift[0], phase_shift[1], basisStates[ell]);

            ZMatrixRMaj norm = new ZMatrixRMaj(1, 1);
            CommonOps_ZDRM.multTransA(basisStates[ell], basisStates[ell], norm);
            CommonOps_ZDRM.scale(1 / norm.getReal(0, 0), 1, basisStates[ell]);
        }

        createMatrix();
    }

    @Override
    public ZMatrixRMaj[] getBasisStates() {
        return basisStates;
    }

    @Override
    public ZMatrixRMaj getBasisMatrix() {
        return basisMatrix;
    }

    @Override
    public int getDimension() {
        return dimension; // the number of cat basis states
    }

    @Override
    public String toString() {
        return "Cat basis matrix of dimension " + getDimension() + ":\n" + MatrixIO.prettyFormat(getBasisMatrix());
    }

    private static long[] computeFactorials(int n) {
        long[] result = new long[n + 1];
        result[0] = 1;
        for (int i = 1; i <= n; i++) {
            result[i] = result[i - 1] * i;
        }
        return result;
    }

    private static double[] complexPower(double[] z, int n) {
        if (n == 0) {
            return new double[]{1.0, 0.0};
        }

        double r = Math.hypot(z[0], z[1]);
        double theta = Math.atan2(z[1], z[0]);

        double rPowered = Math.pow(r, n);
        double resultReal = rPowered * Math.cos(n * theta);
        double resultImag = rPowered * Math.sin(n * theta);

        return new double[]{resultReal, resultImag};
    }

    public static double[] expI(double theta) {
        return new double[]{Math.cos(theta), Math.sin(theta)};
    }

}

File: qumode/src/quantum_knapsack/basis/cat_basis.py
from copy import deepcopy
from typing import List

import numpy as np

from src.quantum_knapsack.matrix import ColumnMatrix
from .qumode_basis import QumodeBasis
from .standard_basis import StandardBasis


class CatBasis(QumodeBasis):
    """Cat state basis implementation for quantum modes.

    This class implements the cat state basis, which consists of superpositions
    of coherent states with specific phase relationships.

    Attributes:
        _dimension (int): Dimension of the basis space
        _N (int): Fock space dimension for encoding
        _alpha (complex): Complex amplitude parameter
        _factorials (List[int]): Precomputed factorials for efficiency
        _fock_basis (StandardBasis): Standard basis
        _coherent_states (List[np.ndarray]): Generated coherent states
    """

    def __init__(self, dimension: int, n: int, alpha: complex) -> None:
        """Initialize a cat state basis.

        Args:
            dimension (int): Dimension of the basis space
            n (int): Fock space dimension for encoding
            alpha (complex): Complex amplitude parameter

        Raises:
            ValueError: If dimension or n is less than 1
        """
        if dimension < 1 or n < 1:
            raise ValueError("Dimension and n must be positive integers")

        super().__init__()
        self._dimension: int = dimension
        self._N: int = n
        self._alpha: complex = alpha

        # Initialize components
        self._factorials: List[int] = self._compute_factorials(n)
        self._fock_basis: StandardBasis = StandardBasis(dimension)
        self._coherent_states: List[np.ndarray] = self._generate_coherent_states(n)

        self._create_basis()
        self._create_ladder_operators()

    def _coherent_state(self, k: int) -> np.ndarray:
        """Generate a coherent state for given index k.

        Args:
            k (int): Index of the coherent state

        Returns:
            np.ndarray: Generated coherent state vector
        """
        coherent_state: np.ndarray = np.zeros((self._N, 1), dtype=complex)

        for n in range(self._N):
            temp: np.ndarray = self._fock_basis.get_basis_state(n)

            # Calculate coefficient
            coeff: np.complex128 = np.exp(2j * np.pi * n * k / self._N)
            coeff *= (self._alpha ** n) / np.sqrt(self._factorials[n])

            coherent_state += coeff * temp

        # Apply normalization factor
        coherent_state *= np.exp(-abs(self._alpha) ** 2 / 2)
        return coherent_state

    def _generate_coherent_states(self, n: int) -> List[np.ndarray]:
        """Generate all coherent states for the basis.

        Args:
            n (int): Number of coherent states to generate

        Returns:
            List[np.ndarray]: List of coherent state vectors
        """
        return [self._coherent_state(k) for k in range(n)]

    def _cat_state(self, ell: int) -> np.ndarray:
        """Generate a cat state for given index ell.

        Args:
            ell (int): Index of the cat state

        Returns:
            np.ndarray: Generated cat state vector
        """
        cat_state: np.ndarray = np.zeros((self._N, 1), dtype=complex)

        for k in range(self._N):
            temp: np.ndarray = deepcopy(self._coherent_states[k])
            coeff: np.complex128 = np.exp(-2j * np.pi * ell * k / self._N)
            cat_state += coeff * temp

        return ColumnMatrix.normalise(cat_state)

    def _create_basis(self) -> None:
        """Create the cat state basis vectors."""
        self._basis_states = [self._cat_state(ell) for ell in range(self._N)]
        self._create_matrix()

    @staticmethod
    def _compute_factorials(n: int) -> List[int]:
        """Precompute factorials up to n for efficiency.

        Args:
            n (int): Maximum number to compute factorial for

        Returns:
            List[int]: List of factorials from 0! to n!
        """
        result: List[int] = [1] * (n + 1)
        for i in range(1, n + 1):
            result[i] = result[i - 1] * i
        return result
